File: src/domain/ingredient.ts
/**
 * Ingredient (食材) ドメイン型。
 *
 * - 食材は地元 (LocaleId) に紐づく
 * - 旬 (Season) は複数許容: 例 ["winter", "spring"]
 * - searchQuery は楽天ふるさと納税 API 検索用 (Slice 5)
 */

import type { LocaleId } from './locale';

export type IngredientId = string;

export type Season = 'spring' | 'summer' | 'autumn' | 'winter' | 'all-year';

export type IngredientCategory = 'vegetable' | 'seafood' | 'cheese' | 'grain' | 'meat' | 'fruit';

export type Ingredient = {
  id: IngredientId;
  localeId: LocaleId;
  name: string;
  searchQuery?: string;
  category: IngredientCategory;
  seasons: Season[];
  story?: string;
};

export const SEASONS: readonly Season[] = [
  'spring',
  'summer',
  'autumn',
  'winter',
  'all-year',
] as const;

export const INGREDIENT_CATEGORIES: readonly IngredientCategory[] = [
  'vegetable',
  'seafood',
  'cheese',
  'grain',
  'meat',
  'fruit',
] as const;

export function isSeason(value: unknown): value is Season {
  return typeof value === 'string' && (SEASONS as readonly string[]).includes(value);
}

export function isIngredientCategory(value: unknown): value is IngredientCategory {
  return typeof value === 'string' && (INGREDIENT_CATEGORIES as readonly string[]).includes(value);
}

/**
 * 指定した季節に該当する食材かを判定する。
 * "all-year" は常に true を返す。
 */
export function isInSeason(ingredient: Ingredient, season: Season): boolean {
  return ingredient.seasons.includes('all-year') || ingredient.seasons.includes(season);
}

File: agent/src/makelocal_agent/domain/ingredient.py
"""Ingredient ドメイン型 — TypeScript の src/domain/ingredient.ts と semantic 同期。

- Season は 5 値 (spring/summer/autumn/winter/all-year)
- IngredientCategory は 6 値
- is_in_season は all-year を常に true として扱う
"""

from __future__ import annotations

from typing import Literal, get_args

from pydantic import BaseModel, Field

from .locale import LocaleId

type IngredientId = str

Season = Literal["spring", "summer", "autumn", "winter", "all-year"]

IngredientCategory = Literal[
    "vegetable",
    "seafood",
    "cheese",
    "grain",
    "meat",
    "fruit",
]

SEASONS: tuple[Season, ...] = get_args(Season)
INGREDIENT_CATEGORIES: tuple[IngredientCategory, ...] = get_args(IngredientCategory)


class Ingredient(BaseModel):
    """食材 1 件 (Tap2 の選択単位)。"""

    id: IngredientId = Field(min_length=1)
    localeId: LocaleId = Field(min_length=1)
    name: str = Field(min_length=1)
    searchQuery: str | None = None
    category: IngredientCategory
    seasons: list[Season] = Field(min_length=1)
    story: str | None = None


def is_season(value: object) -> bool:
    """値が Season 列挙のいずれかかを判定する。"""
    return isinstance(value, str) and value in SEASONS


def is_ingredient_category(value: object) -> bool:
    """値が IngredientCategory 列挙のいずれかかを判定する。"""
    return isinstance(value, str) and value in INGREDIENT_CATEGORIES


def is_in_season(ingredient: Ingredient, season: Season) -> bool:
    """all-year は常に true、それ以外は seasons リストに含まれているか。

    TS 側 isInSeason() と同じ挙動。
    """
    return "all-year" in ingredient.seasons or season in ingredient.seasons
